packages/story-editor/src/components/panels/design/layer/layerList.js

```diff
@@ -53,15 +53,43 @@ const LayerSeparator = styled(ReorderableSeparator)`
   padding: ${LAYER_HEIGHT / 2}px 0;
 `;
 
+const ReorderableLayer = memo(function ReorderableLayer({
+  id,
+  position,
+  handleStartReordering,
+}) {
+  const element = useStory((v) =>
+    v.state.currentPage?.elements.find((el) => el.id === id)
+  );
+  return element ? (
+    <Fragment key={id}>
+      <LayerSeparator position={position + 1} />
+      <ReorderableItem
+        position={position}
+        onStartReordering={handleStartReordering(id)}
+        disabled={element.isBackground}
+      >
+        <Layer element={element} />
+      </ReorderableItem>
+    </Fragment>
+  ) : null;
+});
+
+ReorderableLayer.propTypes = {
+  id: PropTypes.string.isRequired,
+  position: PropTypes.number.isRequired,
+  handleStartReordering: PropTypes.func.isRequired,
+};
+
 function LayerPanel({ layers }) {
-  const { arrangeElement, setSelectedElementsById } = useStory((state) => ({
-    arrangeElement: state.actions.arrangeElement,
-    setSelectedElementsById: state.actions.setSelectedElementsById,
-  }));
+  const arrangeElement = useStory((v) => v.actions.arrangeElement);
+  const setSelectedElementsById = useStory(
```

Comment:
What‘s the benefit of splitting this into two function calls? Just because the lib‘s usage example does it like this?

Reply:
yea so the library is recommending this usage cus of the referential equality blurb I mentioned above.

if the selector returns a new object, this hook will force a re-render of this component every time the provider this context is coming from re-renders. Basically by returning a new object from the selector we're opting out of all the performance benefits `useContextSelector` provides.

Reply:
We're returning objects from the selectors **a lot**, basically everywhere, and I have never really seen issues with unnecessary re-renders because of it. Do you perhaps have a small repro case to demonstrate that it actually doesn't work as intended?

If returning objects like this does not work as intended, we would need to do a massive overhaul of the code base to enforce consistent usage in places with improper use.

Reply:
@littlemilkstudio Thanks a lot for the sandboxes! That's very helpful for reasoning about.

Now, your examples demonstrate the effect when using plain `use-context-selector`, but we're using this "patched" version which uses `shallowEqual` by default:

https://github.com/google/web-stories-wp/blob/2d0a090015349a786e676864c5652df5a134893a/packages/react/src/useContextSelector.js

See https://codesandbox.io/s/nice-alex-k5m9l and how there are no re-renders anymore with this patched version.

Reply:
Ah thanks for pointing out that we're using a patched version. Knowing that, we don't need to separate out all the pieces of context value we're grabbing.

I think you may have forgot to save your sandbox? I'm still seeing the same one I sent you